AutoRest/Generators/Extensions/Extensions/ParameterGroupExtensionHelper.cs

```diff
@@ -0,0 +1,204 @@
+                        parameterGroupName = methodGroupName + "-" + methodName + "-" + postfix;
+                    }
+                    else
+                    {
+                        parameterGroupName = specifiedGroupName;
+                    }
+                    return parameterGroupName;
+                }
+            }
+
+            return null;
+        }
+
+        private static IEnumerable<string> ExtractParameterGroupNames(Method method)
+        {
+            return method.Parameters.Select(p => GetParameterGroupName(method.Group, method.Name, p)).Where(name => !string.IsNullOrEmpty(name)).Distinct();
+        }
+
+        private static IEnumerable<ParameterGroup> ExtractParameterGroups(Method method)
+        {
+            IEnumerable<string> parameterGroupNames = ExtractParameterGroupNames(method);
+
+            return parameterGroupNames.Select(parameterGroupName => BuildParameterGroup(parameterGroupName, method));
+        }
+
+        private static IEnumerable<Method> GetMethodsUsingParameterGroup(IEnumerable<Method> methods, ParameterGroup parameterGroup)
+        {
+            return methods.Where(m => ExtractParameterGroupNames(m).Contains(parameterGroup.Name));
+        }
+
+        /// <summary>
+        /// Adds the parameter groups to operation parameters.
+        /// </summary>
+        /// <param name="serviceClient"></param>
+        public static void AddParameterGroups(ServiceClient serviceClient)
+        {
+            if (serviceClient == null)
+            {
+                throw new ArgumentNullException("serviceClient");
+            }
+
+            HashSet<CompositeType> generatedParameterGroups = new HashSet<CompositeType>();
+
+            foreach (Method method in serviceClient.Methods)
+            {
+                //Copy out flattening transformations as they should be the last
+                List<ParameterTransformation> flatteningTransformations = method.InputParameterTransformation.ToList();
+                method.InputParameterTransformation.Clear();
+
+                //This group name is normalized by each languages code generator later, so it need not happen here.
+
+                IEnumerable<ParameterGroup> parameterGroups = ExtractParameterGroups(method);
+
+                List<Parameter> parametersToAddToMethod = new List<Parameter>();
+                List<Parameter> parametersToRemoveFromMethod = new List<Parameter>();
+
+                foreach (ParameterGroup parameterGroup in parameterGroups)
+                {
+                    CompositeType parameterGroupType =
+                        generatedParameterGroups.FirstOrDefault(item => item.Name == parameterGroup.Name);
+
+                    IEnumerable<Method> methodsWhichUseGroup = GetMethodsUsingParameterGroup(serviceClient.Methods, parameterGroup);
+
+                    string methodList = string.Join(", ", methodsWhichUseGroup.Select(m => string.Format("{0}_{1}", m.Group, m.Name)));
```

Comment:
Given the Batch scenario where most types are used for a single operation, the "for the following operations: Foo." reads awkwardly to me.  We could produce more natural text at the expense of a small helper function:

static string OperationsText(IList<string> ops)
{
    if (ops.Count == 1) return ops[0] + " operation";
    if (ops.Count == 2) return ops[0] + " and " + ops[1] + " operations";
    return String.Join(", ", ops.Take(ops.Count - 1)) + ", and " + ops.Last() + " operations";
}